### PR TITLE
Customer-specific webhook configuration

### PR DESCRIPTION
**Pull request by Blokh**
Implementation of definition callback logic for events in the definition

### CODE REVIEW

File: services/workflows-service/src/collection-flow/collection-flow.module.ts [hunk 1]
```diff
@@ -58,9 +58,6 @@ import { SalesforceIntegrationRepository } from '@/salesforce/salesforce-integra
     HookCallbackHandlerService,
     FileService,
     WorkflowEventEmitterService,
-    DocumentChangedWebhookCaller,
-    WorkflowCompletedWebhookCaller,
-    WorkflowStateChangedWebhookCaller,
```
Comment (MatanYadaev) [inline services/workflows-service/src/collection-flow/collection-flow.module.ts]: This caused the webhooks to be dispatched twice, because it was imported here and only in WorkflowModule.

File: services/workflows-service/.env.example [hunk 2]
```diff
@@ -16,7 +16,6 @@ API_KEY=secret
 NODE_ENV=local
 SENTRY_DSN=
 WEBHOOK_SECRET=webhook_secret
```
Comment (alonp99) [inline services/workflows-service/.env.example:18]: We don't need secret either

File: services/workflows-service/scripts/seed.ts [hunk 3]
```diff
@@ -92,12 +94,14 @@ async function seed(bcryptSalt: string | number) {
...
     env.API_KEY,
     'https://assets-global.website-files.com/62827cf4fe5eb528708511d4/645511cb3d3dd84ee28fe04d_CyberAgent.svg',
+    `webhook-shared-secret-${env.API_KEY}`,
   );
   const customer2 = await createCustomer(
     client,
     '2',
     `${env.API_KEY}2`,
     'https://assets-global.website-files.com/62827cf4fe5eb528708511d4/645d26f285bd18467470e7cd_zenhub-logo.svg',
+    `webhook-shared-secret-${env.API_KEY}2`,
```
Comment (alonp99) [inline services/workflows-service/scripts/seed.ts:108]: Can the customer set a general webhook address?

File: services/workflows-service/src/collection-flow/collection-flow.module.ts (hunk 1 above)
Comment (alonp99) [reply services/workflows-service/src/collection-flow/collection-flow.module.ts]: Nice catch

File: services/workflows-service/.env.example (hunk 2 above)
Comment (MatanYadaev) [reply services/workflows-service/.env.example:18]: It's for backward compatiable, to not break Fido in production. They will keep using `Authorization` header for a while, but new customers will use `X-HMAC-SIGNATURE`

File: services/workflows-service/scripts/seed.ts (hunk 3 above)
Comment (MatanYadaev) [reply services/workflows-service/scripts/seed.ts:108]: WDYM? Can you explain please?

File: services/workflows-service/src/events/workflow-completed-webhook-caller.ts [hunk 4]
```diff
@@ -73,28 +90,29 @@ export class WorkflowCompletedWebhookCaller {
...
-          correlationId,
-          environment,
-          data: {
-            ...restRuntimeData.context,
-          },
+      };
+
+      const res = await this.#__axios.post(url, payload, {
+        headers: {
+          'X-Authorization': env.WEBHOOK_SECRET,
```
Comment (alonp99) [inline services/workflows-service/src/events/workflow-completed-webhook-caller.ts:114]: Now i understand its for backward comptablity

Comment (alonp99) [reply services/workflows-service/src/events/workflow-completed-webhook-caller.ts:114]: Leaving comments will be appriceated

**Review by alonp99 — APPROVED** [review]

File: services/workflows-service/src/common/middlewares/user-session-audit.middleware.intg.test.ts [hunk 5]
```diff
@@ -62,7 +62,13 @@ describe('UserSessionAuditMiddleware', () => {
     middleware = new UserSessionAuditMiddleware(app.get(AppLoggerService), app.get(UserService));
     userService = app.get(UserService);
     const prismaService = app.get(PrismaService);
-    const customer = await createCustomer(prismaService, String(Date.now()), 'secret', '');
+    const customer = await createCustomer(
+      prismaService,
+      String(Date.now()),
+      'secret',
+      '',
```
Comment (tomer-shvadron) [inline services/workflows-service/src/common/middlewares/user-session-audit.middleware.intg.test.ts:69]: What does this empty string mean? We should change it to an object parameter with named properties

File: services/workflows-service/src/common/utils/sign/sign.unit.test.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+      // Act
+      const signature1 = sign({ payload, key: key1 });
+      const signature2 = sign({ payload, key: key2 });
+
+      // Assert
+      expect(signature1).not.toBe(signature2);
+    });
+
+    test('When signing different payloads with the same key, it should return different signatures', () => {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/common/utils/sign/sign.unit.test.ts:55]: I don't think this test is necessary

**Review by alonp99 — APPROVED** [review]

File: services/workflows-service/src/events/document-changed-webhook-caller.ts [hunk 7]
```diff
@@ -75,48 +80,69 @@ export class DocumentChangedWebhookCaller {
...
+        workflowRuntimeId: data.updatedRuntimeData.id,
+        ballerineEntityId: data.entityId,
+        correlationId: data.correlationId,
+        environment,
+        data: data.updatedRuntimeData.context,
+      };
+
+      const res = await this.#__axios.post(url, payload, {
+        headers: {
+          'X-Authorization': env.WEBHOOK_SECRET,
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/document-changed-webhook-caller.ts:142]: We should extract header names to a shared constant (either in `@ballerine/common` if needed or in the `workflows-service` package)

File: services/workflows-service/src/events/document-changed-webhook-caller.ts [hunk 8]
```diff
@@ -19,8 +23,9 @@ export class DocumentChangedWebhookCaller {
   constructor(
     private httpService: HttpService,
     private readonly configService: ConfigService,
-    private workflowEventEmitter: WorkflowEventEmitterService,
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/document-changed-webhook-caller.ts:22]: Why did you remove the private in this line?

File: services/workflows-service/src/workflow/schemas/zod-schemas.ts [hunk 9]
```diff
@@ -1,12 +1,14 @@
...
-  .object({
-    type: z.enum(['webhook', 'slack']),
-    url: z.string().url().optional(),
-    events: z.array(z.string()),
-  })
-  .strict();
+export const SubscriptionSchema = z.discriminatedUnion('type', [
+  z
+    .object({
+      type: z.literal('webhook'),
```
Comment (tomer-shvadron) [inline services/workflows-service/src/workflow/schemas/zod-schemas.ts:6]: Why do we need `type` if it's always `"webhook"`?

File: services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts [hunk 10]
```diff
@@ -41,45 +46,56 @@ export class WorkflowStateChangedWebhookCaller {
...
-          workflowResolvedAt: data.runtimeData.resolvedAt,
-          workflowDefinitionId: data.runtimeData.workflowDefinitionId,
-          workflowRuntimeId: data.runtimeData.id,
-          ballerineEntityId: data.entityId,
-          correlationId: data.correlationId,
-          environment,
-          data: data.runtimeData.context,
+      const res = await this.#__axios.post(url, payload, {
+        headers: {
+          'X-Authorization': env.WEBHOOK_SECRET,
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts:95]: Same as previous comment

File: services/workflows-service/src/workflow/workflow-runtime-data.repository.intg.test.ts [hunk 11]
```diff
@@ -79,7 +79,13 @@ describe('#Workflow Runtime Repository Integration Tests', () => {
       PrismaModule,
     ])) as unknown as PrismaService;
 
-    const customer = await createCustomer(prismaService, '1', 'secret', '');
+    const customer = await createCustomer(
+      prismaService,
+      '1',
+      'secret',
+      '',
```
Comment (tomer-shvadron) [inline services/workflows-service/src/workflow/workflow-runtime-data.repository.intg.test.ts:86]: Same as previous comment

File: services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts [hunk 12]
```diff
@@ -9,17 +9,22 @@ import { AxiosInstance } from 'axios';
...
+import { CustomerService } from '@/customer/customer.service';
+import { TAuthenticationConfiguration } from '@/customer/types';
 
 @Injectable()
 export class WorkflowStateChangedWebhookCaller {
   #__axios: AxiosInstance;
 
   constructor(
     private httpService: HttpService,
-    private workflowEventEmitter: WorkflowEventEmitterService,
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts:20]: Same as previous comment

File: services/workflows-service/src/events/document-changed-webhook-caller.ts [hunk 13]
```diff
@@ -75,48 +80,69 @@ export class DocumentChangedWebhookCaller {
...
-    const { id, environment, url, authSecret, apiVersion } = getWebhookInfo(
+    const webhooks = getWebhooks(
       data.updatedRuntimeData.config,
       this.configService.get('NODE_ENV'),
-      this.configService.get('WEBHOOK_URL'),
-      this.configService.get('WEBHOOK_SECRET'),
       'workflow.context.document.changed',
     );
 
-    if (!url) {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/document-changed-webhook-caller.ts:86]: Do we have any compensation for losing this check?

File: services/workflows-service/src/events/workflow-completed-webhook-caller.ts [hunk 14]
```diff
@@ -9,22 +9,27 @@ import { AxiosInstance } from 'axios';
...
+import { CustomerService } from '@/customer/customer.service';
 
 @Injectable()
 export class WorkflowCompletedWebhookCaller {
   #__axios: AxiosInstance;
 
   constructor(
     private httpService: HttpService,
-    private workflowEventEmitter: WorkflowEventEmitterService,
+    workflowEventEmitter: WorkflowEventEmitterService,
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-completed-webhook-caller.ts:28]: Same as previous comment

File: services/workflows-service/src/events/workflow-completed-webhook-caller.ts [hunk 15]
```diff
@@ -48,19 +53,31 @@ export class WorkflowCompletedWebhookCaller {
...
-    const { id, environment, url, authSecret, apiVersion } = getWebhookInfo(
+    const webhooks = getWebhooks(
       data.runtimeData.config,
       this.configService.get('NODE_ENV'),
-      this.configService.get('WEBHOOK_URL'),
-      this.configService.get('WEBHOOK_SECRET'),
       'workflow.completed',
     );
 
-    if (!url) {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-completed-webhook-caller.ts:59]: Same as previous comment

File: services/workflows-service/src/events/document-changed-webhook-caller.ts [hunk 16]
```diff
@@ -75,48 +80,69 @@ export class DocumentChangedWebhookCaller {
...
     data.updatedRuntimeData.context.documents.forEach((doc: any) => {
       delete doc.propertiesSchema;
     });
 
+    const customer = await this.customerService.getByProjectId(data.updatedRuntimeData.projectId);
+
+    const { webhookSharedSecret } =
+      customer.authenticationConfiguration as TAuthenticationConfiguration;
+
+    for (const webhook of webhooks) {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/document-changed-webhook-caller.ts:98]: Any reason not to use `Promise.all([...])` like this:
```
const webhookPromises = webhooks.map((webhook) => 
    this.sendWebhook({ data, newDocumentsByIdentifier, oldDocuments, webhook, webhookSharedSecret })
);

await Promise.all(webhookPromises);
```
This way all webhooks will fire concurrently

File: services/workflows-service/src/events/workflow-completed-webhook-caller.ts [hunk 17]
```diff
@@ -48,19 +53,31 @@ export class WorkflowCompletedWebhookCaller {
...
-      this.logger.log(`No webhook url found for a workflow runtime data with an id of "${id}"`);
-      return;
+    const customer = await this.customerService.getByProjectId(data.runtimeData.projectId);
+
+    const { webhookSharedSecret } =
+      customer.authenticationConfiguration as TAuthenticationConfiguration;
+
+    for (const webhook of webhooks) {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-completed-webhook-caller.ts:67]: Same as previous comment

File: services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts [hunk 18]
```diff
@@ -41,45 +46,56 @@ export class WorkflowStateChangedWebhookCaller {
...
 
-    if (!url) {
-      this.logger.log(`No webhook url found for a workflow runtime data with an id of "${id}"`);
-      return;
+    const customer = await this.customerService.getByProjectId(data.runtimeData.projectId);
+
+    const { webhookSharedSecret } =
+      customer.authenticationConfiguration as TAuthenticationConfiguration;
+
+    for (const webhook of webhooks) {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/events/workflow-state-changed-webhook-caller.ts:60]: Same as previous comment

**Review by tomer-shvadron — APPROVED** [review]
Awesome work! 👍 Please take a look at the comments I left
